### PR TITLE
Fix the trigger for sending a complaint for VC

### PR DESCRIPTION
**Pull request by yontyon**
Now with the new client manager, consider the following scenario:
1. Request X has beeen received by the primary
2. The primary adds X to the pending requests and starts consensus for a pp message that contains X
3. A nonprimary replica R commits the pp message, as it has never got the client request, it has nothing to remove from its pending requests
4. Now, the client retries X and sends X to all replicas.
5. R gets the client request and puts it in its pending requests
6. X will be never removed from R's pending requests, as it will never be committed again

We fix it by having every replica tracking the client requests that received either by the client or in the pre-prepare.
The later solution was decided in a discussion between @ittaiab @HristoStaykov and @yontyon

### CODE REVIEW

**Review by Liorgal28 — APPROVED** [review]
LGTM, Thanks @yontyon !

File: bftengine/src/bftengine/ClientsManager.cpp [hunk 1]
```diff
@@ -336,8 +336,15 @@ Time ClientsManager::infoOfEarliestPendingRequest(std::string& cid) const {
...
+        earliestPendingReqInfo = {MaxTime, std::string()};
+        earliestTime = MaxTime;
+        break;
+      }
       // Don't take into account already committed requests
-      if ((req.second.time != MinTime) && (earliestTime > req.second.time) && (!req.second.committed)) {
+      if ((req.second.time != MinTime) && (earliestTime > req.second.time)) {
         earliestPendingReqInfo = req.second;
         earliestTime = earliestPendingReqInfo.time;
       }
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ClientsManager.cpp:350]: Maybe it worth changing the name of this function to reflect what it actually does now.

**Review by yuliasherman — APPROVED** [review]
LGTM

**Review by f-squirrel — COMMENTED** [review]
Great catch!

**Review by HristoStaykov — APPROVED** [review]
Great catch and fix @yontyon. Thanks!

**Review by andrewjstone — APPROVED** [review]
Great find, fix, and commit message @yontyon!

**Comment by andrewjstone** [comment]
Actually, @yontyon the great description is only in the MR, but not the commit message. Can you possibly amend the commit to include the description?

**Comment by yontyon** [comment]
In a test that we did, it seems to improve the stability of the system (even that we don't know that this bug does not fully explain the VC, as discussed with @HristoStaykov).
@HristoStaykov I think we should merge it anyway, and track the runs to see if the VC storm is happening again.

**Pull request merged**